[jit] fix dce over loops

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22632 [jit] fix dce over loops**

We need re-run the marking pass over loop sub-blocks until they converge
to a fixed point. Thanks to @chillee for catching this bug!

Differential Revision: [D16184469](https://our.internmc.facebook.com/intern/diff/D16184469)

suo committed:
[jit] fix dce over loops

We need re-run the marking pass over loop sub-blocks until they converge
to a fixed point. Thanks to @chillee for catching this bug!

Chillee commented on torch/csrc/jit/passes/dead_code_elimination.cpp:
```diff
@@ -147,15 +183,23 @@ class DeadCodeEliminator {
       auto node = *it;
-      for (auto subBlock : node->blocks()) {
-        mark(subBlock);
+      if (node->kind() == prim::Loop) {
```
Do we need to check for `c10::onnx::loop` here like we do in the other place?

suo replied:
onnx is purely functional and has no aliasing, so we wouldn't need the special behavior

Chillee reviewed (approved):
Otherwise looks pretty straightforward and correct to me.

Krovatkin commented on torch/csrc/jit/passes/dead_code_elimination.cpp:
```diff
@@ -100,7 +100,8 @@ class DeadCodeEliminator {
       // return block. We consider all graph outputs to be "used", so just mark
       // this node normally.
-      return mark(node);
+      mark(node);
```
I don't quite like the fact that `markReturnNode` isn't changed to `bool`. I get that we only call `mark` for the top-level (function return) and in this case we aren't in a loop, so it doesn't matter, and if we are in a loop we make live only body outputs. When we'll get early returns we might revisiting this and make `markReturnNode` to actually return a `bool`

Krovatkin reviewed (approved):
:shipit:

zdevito reviewed (approved):
Seems fine.

suo committed:
Update on "[jit] fix dce over loops"

[jit] fix dce over loops

We need re-run the marking pass over loop sub-blocks until they converge
to a fixed point. Thanks to @chillee for catching this bug!

gh-metadata: pytorch pytorch 22632 gh/suo/79/head
